### PR TITLE
Update tests and framework to instrument all clients recursively

### PR DESCRIPTION
**Pull request by m-nash**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

**Comment by allenjzhang** [comment]
>             return Resource;

unrelated to this PR, but related to our design discussion where Create model differs from Get model, we will have a problem here.

---
Refers to: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmBuilder.cs:56 in b42629b. [](commit_id = b42629b98f1ed7e3e401e9d9a45a25b2adbf82c7, deletion_comment = False)

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmOperation.cs [hunk 1]
```diff
@@ -28,5 +39,28 @@ protected ArmOperation(TOperations syncValue)
...
+        /// Waits for the completion of the long running operations.
+        /// </summary>
+        /// <param name="cancellationToken"> A token to allow the caller to cancel the call to the service. The default value is <see cref="CancellationToken.None" />. </param>
+        /// <returns> A response with the <see cref="ArmOperation{TOperations}"/> operation for this resource. </returns>
+        /// <remarks>
+        /// <see href="https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-longrunning">Details on long running operation object.</see>
+        /// </remarks>
+        public ArmResponse<TOperations> WaitForCompletion(CancellationToken cancellationToken = default)
+        {
+            var pollingInterval = ArmOperationHelpers<TOperations>.DefaultPollingInterval;
```
Comment (allenjzhang) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmOperation.cs:53]: >DefaultPollingInterval [](start = 67, length = 22)

do we have an ADO item on polling interval based on Retry-After AND/OR a method param override? Currently it is hardcode to 1s.

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Azure.ResourceManager.Core.csproj [hunk 2]
```diff
@@ -19,4 +19,8 @@
     <PackageReference Include="System.Linq.Async" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Compile Include="$(AzureCoreSharedSources)ForwardsClientCallsAttribute.cs" Link="Shared\Core\%(RecursiveDir)\%(Filename)%(Extension)" />
```
Comment (allenjzhang) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/Azure.ResourceManager.Core.csproj:23]: Add to \eng\Directory.Build.data.targets.

Comment (m-nash) [reply sdk/resourcemanager/Azure.ResourceManager.Core/src/Azure.ResourceManager.Core.csproj:23]: This isn't added at that level, all other sdk's add these directly into the proj file
https://github.com/Azure/azure-sdk-for-net/search?l=XML&q=ForwardsClientCallsAttribute

File: sdk/core/Azure.Core.TestFramework/src/ManagementInterceptor.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Reflection;
+using System.Threading.Tasks;
+using Castle.DynamicProxy;
+
+namespace Azure.Core.TestFramework
+{
+    internal class ManagementInterceptor : IInterceptor
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/ManagementInterceptor.cs:10]: Tests?

File: sdk/core/Azure.Core.TestFramework/src/InstrumentClientInterceptor.cs [hunk 4]
```diff
@@ -26,14 +26,26 @@ public void Intercept(IInvocation invocation)
...
-                (type.Name.EndsWith("DefaultSubscription")))
+                (invocation.Method.Name.StartsWith("get_") && type.Name.EndsWith("Operations")))
             {
                 invocation.ReturnValue = _testBase.InstrumentClient(type, result, Array.Empty<IInterceptor>());
+                return;
+            }
+
+            if (
+                // Instrument the subscription client that hangs off of the new AzureResouceManagementClient
+                (invocation.Method.Name.EndsWith("DefaultSubscription")) ||
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/InstrumentClientInterceptor.cs:42]: Can we base this logic on the method returned type ?

File: sdk/core/Azure.Core.TestFramework/src/UseSyncMethodsInterceptor.cs [hunk 5]
```diff
@@ -36,6 +36,12 @@ public UseSyncMethodsInterceptor(bool forceSync)
         [DebuggerStepThrough]
         public void Intercept(IInvocation invocation)
         {
+            if (invocation.Method.Name.Contains("WaitForCompletionAsync"))
```
Comment (pakrym) [inline sdk/core/Azure.Core.TestFramework/src/UseSyncMethodsInterceptor.cs:39]: We don't currently instrument data-plane operations and ArmOperation has a method we can forward to. Why the exclude?

**Comment by pakrym** [comment]
LGTM

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Azure.ResourceManager.Core.csproj (hunk 2 above)
Comment (allenjzhang) [reply sdk/resourcemanager/Azure.ResourceManager.Core/src/Azure.ResourceManager.Core.csproj:22]: sure. For now it's ok. Let's simplify the csproj once we get build.props in order.

**Review by allenjzhang — APPROVED** [review]

**Review by pakrym — APPROVED** [review]

**Pull request merged**